windows/gce-gpu-monitoring-cuda.ps1

```diff
@@ -0,0 +1,227 @@
+#>
+ 
+ Param(
+    #configure the interval, default is 5 and configure bigger than 5 to avoid throttling
+    $interval = 10,
+
+    #cuda is not supported yet
+    [switch]$collect_each_core = $false
+)
+
+$nvidia_smi_path = "C:\Program Files\NVIDIA Corporation\NVSMI\nvidia-smi.exe"
+$metadata_server = "http://metadata/computeMetadata/v1/instance/"
+$access_token_command = 'gcloud auth application-default print-access-token'
+
+$gpu_metrics = @{
+    'utilization.gpu'= 'instance/gpu/utilization';
+    'utilization.memory'= 'instance/gpu/memory_utilization';
+    'memory.total'= 'instance/gpu/memory_total';
+    'memory.used'= 'instance/gpu/memory_used';
+    'memory.free'= 'instance/gpu/memory_free';
+    'temperature.gpu'= 'instance/gpu/temperature';
+}
+
+
+function get_metrics{
+    <#
+    Calls the nvidia-smi tool to retrieve usage metrics about the
+    attached GPUs.
+    :return:
+    Dictionary that maps (gpu_type, gpu_bus_id) to metric values.
+    #>
+    
+    $query_params = 'gpu_name,gpu_bus_id,'+(($gpu_metrics.Keys|sort-object) -join ",")
+
+    $result = & $nvidia_smi_path --query-gpu=$query_params --format=csv,nounits,noheader
+
+    $metric_name = $query_params -split ","
+
+    $final = @{}
+    foreach($r in $result -split "\r?\n"){
+        $r = ($r -split ",").trim()
+        for( $i =2;$i -lt $r.length;$i++){
+            $final[$r[0]+","+$r[1]]+=@{$metric_name[$i]=$r[$i]}
+        }
+    }
+
+    return $final
+    
+}
+
+function get_timeseries_entry{
+    param(
+        $metric_time, 
+        $nvidia_metric,
+        $gcp_metric_name,
+        $value,
+        $gpu_bus_id,
+        $gpu_type
+        )
+
+
+   @{'metric'=@{
+        'type'="custom.googleapis.com/$gcp_metric_name"
+        'labels'=@{
+            'gpu_type'= $gpu_type
+            'gpu_bus_id'= $gpu_bus_id
+        }
+        }
+        
+    'resource'=@{
+        'type'='gce_instance'
+        'labels'=@{
+            'project_id'= $project_id
+            'instance_id'=$instance_id
+            'zone'=$zone
+            }
+        }
+
+    'points' = @(
+            @{
+                'interval'=@{
+                    'endTime'=$metric_time
+                    }
+                'value'=@{
+                    'int64Value'= $value
+                    }
+                    
+            }        
+        )
+    }
+}
+
+function check_nvidia_smi{
+    <#
+    Checks if the nvidia-smi tool is installed and if it detects any GPUs.
+    Prints message to stderr in case of errors.
+    =return=
+    True if the nvidia-smi tool is available.
+    #>
+
+    $installed = test-path -Path $nvidia_smi_path
+    if (-not($installed)){
+        write-host -ForegroundColor Red "Couldn't find the nvidia-smi tool. Make sure it's properly installed in one of the directories in $nvidia_smi_path."
+        return $false
+    }
+
+    $result = & $nvidia_smi_path -L 
+    if ($LASTEXITCODE -ne 0){
```

Comment:
The `if ($?){` should capture the execution success/ failure state more accurately than `$LASTEXITCODE`. I also recommend using `Start-Process` with `-Passthrough` to call the Nvidia SMI binary and determine the state using the returned process obejct.

Reply:
I updated the script to use Start-Process.
However, the $? doesn't work for me. $LASTEXITCODE returned the value I want.